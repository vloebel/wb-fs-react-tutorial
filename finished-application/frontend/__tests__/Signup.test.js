import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { MockedProvider } from '@apollo/react-testing';
import Signup, { SIGNUP_MUTATION } from '../components/Signup';
import { CURRENT_USER_QUERY } from '../components/User';
import { fakeUser } from '../lib/testUtils';

const me = fakeUser();
const mocks = [
  // signup mock mutation
  {
    request: {
      query: SIGNUP_MUTATION,
      variables: {
        name: me.name,
        email: me.email,
        password: 'wes',
      },
    },
    result: {
      data: {
        createUser: {
          __typename: 'User',
          id: 'abc123',
          email: me.email,
          name: me.name,
        },
      },
    },
  },
  // current user query mock
  {
    request: { query: CURRENT_USER_QUERY },
    result: { data: { authenticatedItem: me } },
  },
];

describe('<Signup/>', () => {
  it('renders and matches snapshot', async () => {
    const { container } = render(
      <MockedProvider>
        <Signup />
      </MockedProvider>
    );
    expect(container).toMatchSnapshot();
  });

  it('calls the mutation properly', async () => {
    const { container } = render(
      <MockedProvider mocks={mocks}>
        <Signup />
      </MockedProvider>
    );
    await userEvent.type(screen.getByPlaceholderText('name'), me.name);
    await userEvent.type(screen.getByPlaceholderText('email'), me.email);
    await userEvent.type(screen.getByPlaceholderText('password'), 'wes');
    await userEvent.click(screen.getByText('Sign Up!'));
    // loading state
    expect(screen.getByTestId('loading')).toHaveAttribute('aria-busy', 'true');
    expect(screen.getByTestId('loading')).toHaveAttribute('disabled');
    await screen.findByText(`Signed up with ${me.email} â€” Please Sign In now`);
  });
});
