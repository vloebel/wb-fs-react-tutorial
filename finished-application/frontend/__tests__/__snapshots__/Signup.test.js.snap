// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Signup/> renders and matches snapshot 1`] = `
<div>
  <form
    class="Form-sc-1xszr8q-0 ftBRXq"
    method="post"
  >
    <fieldset
      aria-busy="false"
      data-testid="loading"
    >
      <h2>
        Sign Up for An Account
      </h2>
      <label
        for="email"
      >
        Email
        <input
          autocomplete="email"
          name="email"
          placeholder="email"
          type="email"
          value=""
        />
      </label>
      <label
        for="name"
      >
        Name
        <input
          autocomplete="name"
          name="name"
          placeholder="name"
          type="text"
          value=""
        />
      </label>
      <label
        for="password"
      >
        Password
        <input
          autocomplete="new-password"
          name="password"
          placeholder="password"
          type="password"
          value=""
        />
      </label>
      <button
        type="submit"
      >
        Sign Up!
      </button>
    </fieldset>
  </form>
</div>
`;
